go-controller/hybrid-overlay/pkg/controller/node_linux.go

```diff
@@ -53,27 +59,156 @@ func podToCookie(pod *kapi.Pod) string {
 	return nameToCookie(pod.Namespace + "_" + pod.Name)
 }
 
+func (n *NodeController) waitForNamespace(name string) (*kapi.Namespace, error) {
+	var namespaceBackoff = wait.Backoff{Duration: 1 * time.Second, Steps: 7, Factor: 1.5, Jitter: 0.1}
+	var namespace *kapi.Namespace
+	if err := wait.ExponentialBackoff(namespaceBackoff, func() (bool, error) {
+		var err error
+		namespace, err = n.wf.GetNamespace(name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				// Namespace not found; retry
+				return false, nil
+			}
+			klog.Warningf("error getting namespace: %v", err)
+			return false, err
+		}
+		return true, nil
+	}); err != nil {
+		return nil, fmt.Errorf("failed to get namespace object: %v", err)
+	}
+	return namespace, nil
+}
+
 func (n *NodeController) addOrUpdatePod(pod *kapi.Pod) error {
 	podIPs, podMAC, err := getPodDetails(pod, n.nodeName)
 	if err != nil {
 		klog.V(5).Infof("cleaning up hybrid overlay pod %s/%s because %v", pod.Namespace, pod.Name, err)
 		return n.deletePod(pod)
 	}
 
+	namespace, err := n.waitForNamespace(pod.Namespace)
+	if err != nil {
+		return err
+	}
+	hybridOverlayExternalGw := namespace.GetAnnotations()[hotypes.HybridOverlayExternalGw]
+	namespaceVTEP := namespace.GetAnnotations()[hotypes.HybridOverlayVTEP]
 	cookie := podToCookie(pod)
+	if !n.initialized {
+		node, err := n.wf.GetNode(n.nodeName)
+		if err != nil {
+			return fmt.Errorf("hybrid overlay not initialized on %s, and failed to get node data: %v",
+				n.nodeName, err)
+		}
+		if err = n.ensureHybridOverlayBridge(node); err != nil {
+			return fmt.Errorf("failed to ensure hybrid overlay in pod handler: %v", err)
+		}
+	}
+	if len(n.drMAC) == 0 {
```

Comment:
Neither of these things (n.drMAC and n.drIP) will ever be empty if we get this far, because they are set in ensureHybridOverlayBridge()

Reply:
I think they can, because ensureHybridOverlayBridge is really "tryToEnsureHybridOverlayBridge". The function will return a nil if the node does not have the right annotations, with a "try again later" mentality. However, there is a duplicate set of checks for this on line 129. I'll go ahead and remove those.